From a6d0e9d13da23ff2fdfb1bc08dbf0600bd039459 Mon Sep 17 00:00:00 2001
From: Jan Sondhauss <jan.sondhauss@wago.com>
Date: Tue, 2 Apr 2019 10:12:30 +0200
Subject: [PATCH] net: phy: always suspend phy when entering PHY_HALTED

This fixes a bug where the corresponding netlink NEW_LINK message
was not sent when the link was forced (i.e. ANEG disabled).
We now make sure that the __LINK_STATE_NO_CARRIER flag is always
set.

Signet-off-by: Jan Sondhauss <jan.sondhauss@wago.com>
---
 drivers/net/phy/phy.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/drivers/net/phy/phy.c b/drivers/net/phy/phy.c
index eb31d8a..222f180d 100644
--- a/drivers/net/phy/phy.c
+++ b/drivers/net/phy/phy.c
@@ -1094,12 +1094,10 @@ void phy_state_machine(struct work_struct *work)
 						   PHY_INTERRUPT_ENABLED);
 		break;
 	case PHY_HALTED:
-		if (phydev->link) {
-			phydev->link = 0;
-			netif_carrier_off(phydev->attached_dev);
-			phydev->adjust_link(phydev->attached_dev);
-			do_suspend = true;
-		}
+		phydev->link = 0;
+		netif_carrier_off(phydev->attached_dev);
+		phydev->adjust_link(phydev->attached_dev);
+		do_suspend = true;
 		break;
 	case PHY_RESUMING:
 		if (AUTONEG_ENABLE == phydev->autoneg) {
-- 
2.7.4

